wasm/validation.h

```diff
@@ -0,0 +1,47 @@
+// SPDX-FileCopyrightText: 2024 David Zero <zero-one@zer0-one.net>
+//
+// SPDX-License-Identifier: BSD-2-Clause
+
+#ifndef WASM_VALIDATION_H_
+#define WASM_VALIDATION_H_
+
+#include "wasm.h"
+#include "wasm/instructions.h"
+#include "wasm/types.h"
+
+#include <cassert>
+#include <cstdint>
+#include <variant>
+
+namespace wasm::validation {
+
+constexpr bool is_valid(Limits const &l, std::uint32_t k) {
+    if (l.min > k) {
+        return false;
+    }
+
+    if (l.max.has_value()) {
+        if (l.max > k || l.max < l.min) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+constexpr bool is_valid(wasm::instructions::BlockType &bt, Module const &m) {
+    assert(m.type_section.has_value());
+
+    if (auto *t = std::get_if<TypeIdx>(&bt.value)) {
+        if (m.type_section.value().types.size() <= *t) {
```

Comment:
When you know an optional (m.type_section) has a value, prefer unchecked access like `*` or `->`. `.value()` throws an exception or terminates when the optional is empty, whereas the unchecked ones cause UB.